### PR TITLE
return empty sources for non-dart uris in BuildAssetUriResolver

### PR DESCRIPTION
**Pull request by jakemac53**
This allows missing parts to appear in `LibraryElement#parts`.

cc @scheglov

### CODE REVIEW

File: bazel_codegen/test/build_asset_uri_resolver_test.dart [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  });
+
+  test('returns null for dart: uris', () {
+    expect(resolver.resolveAbsolute(Uri.parse('dart:async')), isNull);
+  });
+
+  test('returns an empty source for unknown assets', () {
+    var source = resolver.resolveAbsolute(Uri.parse('package:foo/bar.dart'));
+    expect(source.exists(), isFalse);
+    expect(() => source.contents, throwsA(const TypeMatcher<StateError>()));
```
Comment (scheglov) [inline bazel_codegen/test/build_asset_uri_resolver_test.dart]: There is already `throwsStateError` in `package:test`.

File: bazel_codegen/lib/src/summaries/build_asset_uri_resolver.dart [hunk 2]
```diff
@@ -23,13 +23,17 @@ class BuildAssetUriResolver extends UriResolver {
...
       if (!_knownAssets.containsKey(uri)) {
-        _knownAssets[uri] = AssetSource(asset, await read(asset));
+        _knownAssets[uri] = AssetSource(asset, await read(asset), true);
       }
     }
   }
 
   @override
-  Source resolveAbsolute(Uri uri, [Uri actualUri]) => _knownAssets[uri];
+  Source resolveAbsolute(Uri uri, [Uri actualUri]) => _knownAssets.putIfAbsent(
```
Comment (natebosch) [inline bazel_codegen/lib/src/summaries/build_asset_uri_resolver.dart]: I worry about this returning empty sources for URIs that should be handled by summaries.

This resolver comes [earlier in the list](https://github.com/dart-lang/build/blob/fc63e34c54e2194b00d7c8d524ee1bd71d13a0b3/bazel_codegen/lib/src/summaries/analysis_context.dart#L32-L35) so I'm confused how that isn't already a problem...

Comment (jakemac53) [reply bazel_codegen/lib/src/summaries/build_asset_uri_resolver.dart]: Good catch - I needed to move this to the end of the list.

Comment (jakemac53) [reply bazel_codegen/lib/src/summaries/build_asset_uri_resolver.dart]: The api for that looks a lot different... @scheglov do you know what we should be doing for analysis driver?

Do we need to use `resolver.newFile` with null content or will it handle this better by default?

**Comment by jakemac53** [comment]
> Does the `AnalysisDriver` code path require the same behavior?

The api for that looks a lot different, @scheglov do we need to use `resolver.addFile` with null content or will analysis driver do something better by default here?

**Comment by jakemac53** [comment]
@natebosch lets follow up with the analysis driver changes

**Comment by scheglov** [comment]
I don't see the code about which you are asking the question, not sure how to navigate to it on GitHub. AnalysisDriver needs just URI to file path resolution (and back). The file does not have to exist, if a file does not exist, it is treated as empty.

**Comment by jakemac53** [comment]
@scheglov The new uri resolver is here https://github.com/dart-lang/build/blob/master/bazel_codegen/lib/src/resolver/build_asset_uri_resolver.dart

It will end up returning null for `resolveAbsolute` today but I wasn't sure how to create an empty source since we don't control that any more.

**Comment by scheglov** [comment]
If a URI cannot be resolved to a file path, it will be considered invalid and skipped.
Corresponding import / export / part directives will not be reflected in the element model of the library.

**Comment by jakemac53** [comment]
> If a URI cannot be resolved to a file path, it will be considered invalid and skipped.
> Corresponding import / export / part directives will not be reflected in the element model of the library.

Ok - so back to my original question then, how do we create/return an empty source? We can create a uri for these files, it just isn't obvious how to get the source.

More specifically it looks like today we effectively do `resourceProvider.newFile(path, content)` with the known files, and then `resourceProvider.getFile(path).createSource(uri)`.

If we know we haven't ever called `newFile` before with a path can we still just call `getFile` with it and then subsequently call `createSource` on that object? Or will we get a null back from `getFile` or an exception?

**Comment by scheglov** [comment]
`resourceProvider.getFile(path)` return a handle object.
The files does not have to exist, you will always get a not `null` object.
When you attempt to read it, and it still does not exist, then you will get an exception.
But otherwise you can work with it - ask its path,  create `Source`.

**Pull request merged**